### PR TITLE
feat: expose CollisionData in Collisions component

### PR DESCRIPTION
**Pull request by agg23**
Currently the `Collisions` component only exposes the entities present in a collision, but there is additional useful information available, such as the collision normals. This change exposes the `CollisionData` object in the `Collisions` component.

Additionally, the collision normals were captured as `Vec2`s, dropping the Z component. I swapped to a `Vec3`, zeroing the Z component when in 2D.

### CODE REVIEW

File: core/src/collisions.rs [hunk 1]
```diff
@@ -22,12 +22,22 @@ impl Collisions {
     /// Returns `true` if the collisions contains the specified entity.
     #[must_use]
     pub fn contains(&self, entity: &Entity) -> bool {
-        self.0.contains(entity)
+        self.0.contains_key(entity)
+    }
+
+    /// An iterator visiting all colliding entities and the resulting collision data in arbitrary order.
+    pub fn iter(&self) -> impl Iterator<Item = (Entity, &CollisionData)> + '_ {
+        self.0.iter().map(|(e, d)| (*e, d))
```
Comment (jcornaz) [inline core/src/collisions.rs:30]: That is a breaking change, restore the previous declaration.

File: core/src/events.rs [hunk 2]
```diff
@@ -136,7 +136,7 @@ impl CollisionData {
 
     /// Returns the normal vector pointing toward this entity and away from the other entity
     #[must_use]
-    pub fn normals(&self) -> &[Vec2] {
+    pub fn normals(&self) -> &[Vec3] {
         &self.normals
     }
 }
```
Comment (jcornaz) [inline core/src/events.rs:142]: This is an unrelated change. Please revert and open another PR.

Also, note that this is a breaking change. I understand, it is due to the mistake of having exposed 2d vectors instead of the 3d ones. But nevertheless, we shall try to improve it without breaking the API.

**Review by jcornaz — CHANGES_REQUESTED** [review]
Hi, thanks for the contribution.

Please remove the breaking changes and separate the change of collision data into another PR.

**Comment by jcornaz** [comment]
> @jcornaz what are your thoughts around breaking changes? It seems odd to build a less than idiomatic API (the `iter` breaking change above) to avoid breaking changes. Maybe it's ideal to not use generic method names like `iter` to avoid this problem, but since that's the Rust standard, I'm not really sure what we should do about it.

I don't think there is much to do about it. `iter` sounded like a good name at that time, we can deprecate it (and hide it from the docs) when that will no longer be the case. I don't think that is a problem, software dev is about iterating and improving over time when we find a better solution. Only breaking the API is a problem. As long as we introduce improvements without breaking the API, I am happy.


> I'm assuming your intention is to have
> 
> ```rust
> #[deprecated]
> fn iter(&self) -> impl Iterator<Item = Entity> {}
> fn entities(&self) -> impl Iterator<Item = Entity> {}
> fn collision_data(&self) -> impl Iterator<Item = &CollisionData> {}
> ```

That's a possible way yes. And probably the best I can think of right now. But I don't have a very strong opinion and welcome other suggestions. As long as there is no breaking change :-p

Also add `#[doc(hidden)]` on the deprecated member, as well as a message explaining to call `entities` instead.

> and some fourth option to return the pairs?

I don't think we need an iterator over the pairs. The collision data already contains the entities. Unless you actually need it for your current use-case?

File: core/src/events.rs [hunk 3]
```diff
@@ -33,7 +33,7 @@ pub enum CollisionEvent {
...
 
 /// Collision data concerning one of the two entity that collided
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Reflect)]
```
Comment (jcornaz) [inline core/src/events.rs:36]: Why add these two traits?

Comment (agg23) [reply core/src/events.rs:36]: I believe the `PartialEq` is only there for the tests (so not really needed), but `Collisions` is `Reflect`, so `CollisionData` needs to be `Reflect` as well (unless we make a custom implementation).

**Review by jcornaz — APPROVED** [review]
Thanks :+1:
